OrbitService/ProcessList.h

```diff
@@ -9,20 +9,61 @@
 #include <vector>
 
 #include "OrbitBase/Result.h"
+#include "Process.h"
+#include "absl/container/flat_hash_set.h"
 #include "process.pb.h"
 
 namespace orbit_service {
 
 class ProcessList {
  public:
   [[nodiscard]] ErrorMessageOr<void> Refresh();
-  [[nodiscard]] const std::vector<orbit_grpc_protos::ProcessInfo>& GetProcesses() {
-    return processes_;
+  [[nodiscard]] std::vector<orbit_grpc_protos::ProcessInfo> GetProcesses() const {
+    std::vector<orbit_grpc_protos::ProcessInfo> processes;
+    processes.reserve(processes_.size());
+
+    std::copy(processes_.begin(), processes_.end(), std::back_inserter(processes));
+    return processes;
+  }
+
+  [[nodiscard]] std::optional<const Process*> GetProcessByPid(pid_t pid) const {
+    const auto it = processes_.find(pid);
+    if (it != processes_.end()) {
+      return &(*it);
+    } else {
+      return {};
+    }
   }
 
  private:
-  std::vector<orbit_grpc_protos::ProcessInfo> processes_;
-  std::unordered_map<int32_t, orbit_grpc_protos::ProcessInfo*> processes_map_;
+  struct ProcessHash {
+    using is_transparent = int;
+
+    [[nodiscard]] size_t operator()(const Process& p) const { return (*this)(p.pid()); }
+
+    [[nodiscard]] size_t operator()(pid_t pid) const { return hash_(pid); }
+
+   private:
+    absl::Hash<pid_t> hash_;
+  };
+
+  struct ProcessEq {
+    using is_transparent = int;
+
+    [[nodiscard]] bool operator()(const Process& lhs, const Process& rhs) const {
+      return lhs.pid() == rhs.pid();
+    }
+
+    [[nodiscard]] bool operator()(const pid_t lhs, const Process& rhs) const {
+      return lhs == rhs.pid();
+    }
+
+    [[nodiscard]] bool operator()(const Process& lhs, const pid_t rhs) const {
+      return lhs.pid() == rhs;
+    }
+  };
+
+  absl::flat_hash_set<Process, ProcessHash, ProcessEq> processes_;
```

Comment:
Wouldn't a map from int32_t to orbit_grpc_protos::ProcessInfo be more intuitive? I have to say that I find it unnecessary complicated to have to define hash and equality that work for both Process and pid_t when the access is (or can be) only by pid.

Reply:
Since `ProcessInfo` already contained the PID, it would mean that the PID is stored in the key and in the value-struct and must be kept in sync. For me personally this feels bug-prone, semantically wrong and less intuitive than a set.

Can you explain where you see the complexity? I admit, defining ProcessHash and ProcessEq is verbose, but is it difficult? Maybe I'm misjudging that...

Reply:
It would be great to get some other opinions of that. Obviously I like my code :sweat_smile: , but I'm also happy to comply with the majority. :see_no_evil:

Reply:
I also think map looks better here.

Reply:
Alright. I changed it to a `absl::flat_hash_map<pid_t, Process>`.